File: src/TrackForge/UI/Theme.cs
namespace TrackForge.UI;

/// <summary>
/// Flat dark palette. No gradients, no glows, no emoji, square corners.
///
/// Accent discipline: <see cref="Accent"/> appears only on the active tab underline,
/// primary buttons, the 2px inset bar on a selected or best-match row, running
/// progress bars, and the uppercase section eyebrows in Settings. Nowhere else.
/// </summary>
public static class Theme
{
    public static readonly Color Background = Color.FromArgb(0x0d, 0x0f, 0x12);
    public static readonly Color Surface = Color.FromArgb(0x15, 0x18, 0x1d);
    public static readonly Color SurfaceAlt = Color.FromArgb(0x1b, 0x1f, 0x25);
    public static readonly Color SurfaceHigh = Color.FromArgb(0x23, 0x28, 0x2f);
    public static readonly Color Border = Color.FromArgb(0x2e, 0x34, 0x3d);

    /// <summary>Divider between table rows.</summary>
    public static readonly Color RowDivider = Color.FromArgb(0x19, 0x1d, 0x23);
    /// <summary>Odd table rows, and non-best rows in Find.</summary>
    public static readonly Color RowOdd = Color.FromArgb(0x17, 0x1b, 0x21);

    /// <summary>Top bar, jobs dock, tool block - chrome rather than content.</summary>
    public static readonly Color ChromePanel = Color.FromArgb(0x12, 0x15, 0x1a);
    public static readonly Color ChromeBorder = Color.FromArgb(0x22, 0x27, 0x2f);

    public static readonly Color Text = Color.FromArgb(0xe8, 0xeb, 0xef);
    public static readonly Color TextStrong = Color.FromArgb(0xc1, 0xc8, 0xd1);
    public static readonly Color TextDim = Color.FromArgb(0x92, 0x9c, 0xa8);
    public static readonly Color TextMuted = Color.FromArgb(0x8a, 0x94, 0xa1);
    public static readonly Color TextFaint = Color.FromArgb(0x62, 0x6c, 0x78);
    public static readonly Color TextFainter = Color.FromArgb(0x5b, 0x65, 0x72);
    public static readonly Color TextCount = Color.FromArgb(0x7d, 0x87, 0x94);

    public static readonly Color Accent = Color.FromArgb(0xc8, 0x44, 0x2f);
    public static readonly Color AccentHover = Color.FromArgb(0xde, 0x54, 0x3d);
    public static readonly Color AccentDim = Color.FromArgb(0x78, 0x2a, 0x1e);

    public static readonly Color Good = Color.FromArgb(0x4e, 0xad, 0x7a);
    public static readonly Color Warn = Color.FromArgb(0xd1, 0x9c, 0x3e);
    public static readonly Color Bad = Color.FromArgb(0xc6, 0x54, 0x54);
    public static readonly Color Selection = Color.FromArgb(0x26, 0x2e, 0x38);

    /// <summary>Border of a grab card once its download finished.</summary>
    public static readonly Color DoneBorder = Color.FromArgb(0x2f, 0x4a, 0x3b);
    /// <summary>Wave-mark bars in the artwork placeholder.</summary>
    public static readonly Color WaveBar = Color.FromArgb(0x3a, 0x42, 0x50);

    // ---- type -------------------------------------------------------------
    // Segoe UI throughout; Consolas for every numeric and path, which is what
    // makes the table columns line up.

    public static readonly Font Body = new("Segoe UI", 9f);                       // 12px 400
    public static readonly Font Emphasis = new("Segoe UI Semibold", 9f);          // 12px 600
    public static readonly Font Secondary = new("Segoe UI", 8.25f);               // 11px 400
    public static readonly Font Eyebrow = new("Segoe UI Semibold", 7.5f);         // 10px 600
    public static readonly Font InspectorTitle = new("Segoe UI Semibold", 10.5f); // 14px 600

    public static readonly Font Numeric = new("Consolas", 9f);
    public static readonly Font NumericSmall = new("Consolas", 8.25f);

    // Kept so older call sites keep compiling.
    public static readonly Font UI = Body;
    public static readonly Font UIBold = Emphasis;
    public static readonly Font Small = Secondary;
    public static readonly Font Mono = Numeric;
    public static readonly Font Heading = InspectorTitle;

    // ---- metrics ----------------------------------------------------------

    public const int Gap = 6;
    public const int Pad = 10;
    public const int FieldHeight = 24;
    public const int RowHeight = 26;
    public const int HeaderHeight = 24;
    public const int ButtonHeight = 24;
    public const int PrimaryButtonHeight = 26;
    public const int TitleBarHeight = 32;
    public const int TopBarHeight = 38;
    public const int JobsDockWidth = 320;
    public const int GrabArtSize = 88;

    /// <summary>Uppercase section label: 10px 600, letter-spaced, accent.</summary>
    public static void DrawEyebrow(Graphics g, string text, Rectangle bounds)
    {
        var spaced = string.Join(" ", text.ToUpperInvariant().ToCharArray());
        TextRenderer.DrawText(g, spaced, Eyebrow, bounds, Accent,
            TextFormatFlags.Left | TextFormatFlags.VerticalCenter | TextFormatFlags.NoPrefix);
    }

    public static void Apply(Control root)
    {
        foreach (Control c in root.Controls)
        {
            switch (c)
            {
                case TextBox tb:
                    tb.BackColor = SurfaceAlt;
                    tb.ForeColor = Text;
                    tb.BorderStyle = BorderStyle.FixedSingle;
                    break;
                case ComboBox cb:
                    cb.BackColor = SurfaceAlt;
                    cb.ForeColor = Text;
                    cb.FlatStyle = FlatStyle.Flat;
                    break;
                case CheckBox chk:
                    chk.ForeColor = Text;
                    chk.BackColor = Color.Transparent;
                    chk.FlatStyle = FlatStyle.Flat;
                    break;
                case Label lbl:
                    lbl.ForeColor = TextDim;
                    lbl.BackColor = Color.Transparent;
                    break;
                case Panel p:
                    if (p.BackColor == SystemColors.Control) p.BackColor = Surface;
                    break;
            }
            if (c.HasChildren) Apply(c);
        }
    }
}

File: macOS/Sources/Theme.swift
import SwiftUI

/// Flat dark palette, carried over from the Windows build.
/// No gradients, no glows, no emoji, square corners.
///
/// Accent discipline: `accent` appears only on the active tab underline, primary
/// buttons, the 2px inset bar on a selected or best-match row, running progress
/// bars, and the uppercase section eyebrows in Settings. Nowhere else.
enum Theme {
    // ---- colour -----------------------------------------------------------

    static let background  = hex(0x0d0f12)
    static let surface     = hex(0x15181d)
    static let surfaceAlt  = hex(0x1b1f25)
    static let surfaceHigh = hex(0x23282f)
    static let border      = hex(0x2e343d)

    /// Divider between table rows.
    static let rowDivider = hex(0x191d23)
    /// Odd table rows, and non-best rows in Find.
    static let rowOdd = hex(0x171b21)

    /// Top bar, jobs dock, tool block — chrome rather than content.
    static let chromePanel  = hex(0x12151a)
    static let chromeBorder = hex(0x22272f)

    static let text       = hex(0xe8ebef)
    static let textStrong = hex(0xc1c8d1)
    static let textDim    = hex(0x929ca8)
    static let textMuted  = hex(0x8a94a1)
    static let textFaint  = hex(0x626c78)
    static let textFainter = hex(0x5b6572)
    static let textCount  = hex(0x7d8794)

    static let accent      = hex(0xc8442f)
    static let accentHover = hex(0xde543d)
    static let accentDim   = hex(0x782a1e)

    static let good = hex(0x4ead7a)
    static let warn = hex(0xd19c3e)
    static let bad  = hex(0xc65454)
    static let selection = hex(0x262e38)

    /// Border of a grab card once its download finished.
    static let doneBorder = hex(0x2f4a3b)
    /// Wave-mark bars in the artwork placeholder.
    static let waveBar = hex(0x3a4250)

    // ---- type -------------------------------------------------------------
    // The system UI face throughout; a monospaced face for every numeric and
    // path, which is what makes the table columns line up.

    static let body      = Font.system(size: 12)
    static let emphasis  = Font.system(size: 12, weight: .semibold)
    static let secondary = Font.system(size: 11)
    static let eyebrow   = Font.system(size: 10, weight: .semibold)
    static let inspectorTitle = Font.system(size: 14, weight: .semibold)

    static let numeric      = Font.system(size: 12, design: .monospaced)
    static let numericSmall = Font.system(size: 11, design: .monospaced)

    // ---- metrics ----------------------------------------------------------

    static let gap: CGFloat = 6
    static let pad: CGFloat = 10
    static let fieldHeight: CGFloat = 24
    static let rowHeight: CGFloat = 26
    static let headerHeight: CGFloat = 24
    static let buttonHeight: CGFloat = 24
    static let primaryButtonHeight: CGFloat = 26
    static let topBarHeight: CGFloat = 38
    static let jobsDockWidth: CGFloat = 320
    static let grabArtSize: CGFloat = 88

    static func hex(_ value: UInt32) -> Color {
        Color(
            .sRGB,
            red:   Double((value >> 16) & 0xff) / 255,
            green: Double((value >> 8) & 0xff) / 255,
            blue:  Double(value & 0xff) / 255,
            opacity: 1
        )
    }
}

/// Uppercase section label: 10px 600, letter-spaced, accent.
struct Eyebrow: View {
    let text: String
    init(_ text: String) { self.text = text }

    var body: some View {
        Text(text.uppercased())
            .font(Theme.eyebrow)
            .tracking(1.6)
            .foregroundColor(Theme.accent)
    }
}
